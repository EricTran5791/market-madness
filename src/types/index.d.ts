declare module '*.json' {
  /* tslint:disable no-any */
  const value: any;
  export default value;
}
