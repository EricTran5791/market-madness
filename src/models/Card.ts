import { types } from 'mobx-state-tree';
import {
  CardEffectUnion,
  CardEffectKind,
  BasicCardEffectSnapshotType,
  CardEffectCategory,
  InteractiveCardEffectSnapshotType,
  InteractiveCardEffectCategory,
} from './CardEffect';

export enum CardCategory {
  Action = 'Action',
  Consumable = 'Consumable',
  Item = 'Item',
  Money = 'Money',
  NPC = 'NPC',
}

export enum CardSubcategory {
  Food = 'Food',
}

export const Card = types
  .model({
    id: types.identifier(types.string),
    name: types.string,
    category: types.enumeration(
      'Category',
      Object.keys(CardCategory).map(key => CardCategory[key])
    ),
    subcategories: types.optional(
      types.array(
        types.enumeration(
          'Subcategory',
          Object.keys(CardSubcategory).map(key => CardSubcategory[key])
        )
      ),
      []
    ),
    description: types.optional(types.string, ''),
    cost: types.optional(types.number, 0),
    money: types.optional(types.number, 0),
    health: types.optional(types.number, 0),
    isPlayed: types.optional(types.boolean, false),
    effects: types.optional(types.array(CardEffectUnion), []),
  })
  .actions(self => ({
    afterCreate() {
      if (self.description.length > 0 || self.effects.length === 0) {
        return;
      }

      // Auto generate descriptions
      if (self.category === CardCategory.NPC) {
        self.description = 'Defeat: ';
      }
      self.description += self.effects
        .map(effect => {
          if (effect.kind === CardEffectKind.Basic) {
            const {
              category,
              value,
              gainedCardId,
            }: BasicCardEffectSnapshotType = effect;
            switch (category) {
              case CardEffectCategory.Draw:
                return `Draw ${value} card${value > 1 ? 's' : ''}`;
              case CardEffectCategory.GainAttack:
                return `+${value} Attack`;
              case CardEffectCategory.GainCardToDiscardPile:
                // The RegEx converts the camel case id to a spaced and capitalized name
                // TODO: There was a circular dependency when cardGenerator was imported
                // due to cardLibrary also importing CardCategory from Card
                return `Gain ${value} ${gainedCardId
                  .replace(/([A-Z])/g, ' $1')
                  .replace(/^./, (str: string) => {
                    return str.toUpperCase();
                  })}`;
              case CardEffectCategory.GainCardToHand:
                // The RegEx converts the camel case id to a spaced and capitalized name
                // TODO: There was a circular dependency when cardGenerator was imported
                // due to cardLibrary also importing CardCategory from Card
                return `Add ${value} ${gainedCardId
                  .replace(/([A-Z])/g, ' $1')
                  .replace(/^./, (str: string) => {
                    return str.toUpperCase();
                  })} to your hand`;
              case CardEffectCategory.GainMoney:
                return `+${value} Money`;
              case CardEffectCategory.Heal:
                return `Heal ${value}`;
              case CardEffectCategory.IncreaseMaxHealth:
                return `Increase max health by ${value}`;
              case CardEffectCategory.TrashSelf:
                return `Trash this card`;
              default:
                return '';
            }
          }
          if (effect.kind === CardEffectKind.Interactive) {
            const {
              category,
              numCardsToResolve,
            }: InteractiveCardEffectSnapshotType = effect;
            switch (category) {
              case InteractiveCardEffectCategory.OptionalTrash:
                return `Trash up to ${numCardsToResolve} cards`;
              case InteractiveCardEffectCategory.MandatoryDiscard:
                return `Discard ${numCardsToResolve} cards`;
              case InteractiveCardEffectCategory.MandatoryTrash:
                return `Trash ${numCardsToResolve} cards`;
              default:
                return '';
            }
          }
          return '';
        })
        .join(', ');
    },
    /** Reset a card's isPlayed status every time it is detached from its direct parent. */
    beforeDetach() {
      self.isPlayed = false;
    },
  }));

export type CardModelType = typeof Card.Type;
export type CardModelSnapshotType = typeof Card.SnapshotType;

export const CardStack = types
  .model({
    cards: types.array(Card),
  })
  .views(self => ({
    get totalCards(): number {
      return self.cards.length;
    },
    get totalMoney() {
      return self.cards
        .map(card => card.money)
        .reduce((sum, currentValue) => sum + currentValue, 0);
    },
    get unplayedCards() {
      return self.cards.filter(card => !card.isPlayed);
    },
  }))
  .actions(self => ({
    add(card: CardModelType) {
      self.cards.push(card);
    },
    remove(card: CardModelType) {
      self.cards.remove(card);
    },
  }));

export type CardStackModelType = typeof CardStack.Type;
