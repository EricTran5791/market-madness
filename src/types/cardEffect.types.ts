export enum CardEffectKind {
  Basic = 'Basic',
  Interactive = 'Interactive',
}

export enum CardEffectCategory {
  Draw = 'Draw',
  GainAttack = 'Gain Attack',
  GainCardToDiscardPile = 'Gain Card to Discard Pile',
  GainCardToHand = 'Gain Card to Hand',
  GainMoney = 'Gain Money',
  Heal = 'Heal',
  IncreaseMaxHealth = 'Increase Max Health',
  ShuffleCardToDeck = 'Shuffle Card to Deck',
  TopdeckCardToOpposingShopDecks = 'Topdeck Card to Opposing Shop Decks',
  TrashSelf = 'Trash Self',
}

export enum InteractiveCardEffectCategory {
  Discard = 'Discard',
  Trash = 'Trash',
}

export enum InteractiveCardEffectResolveType {
  Mandatory = 'Mandatory',
  Optional = 'Optional',
}

export type GainedCard = {
  id: string;
  name: string;
};

export interface BasicCardEffect {
  kind: CardEffectKind.Basic;
  category: CardEffectCategory;
  value: number;
  gainedCard?: GainedCard;
}

export interface InteractiveCardEffect {
  kind: CardEffectKind.Interactive;
  category: InteractiveCardEffectCategory;
  resolveType: InteractiveCardEffectResolveType;
  numPlaysToResolve?: number;
}

export type CardEffect = BasicCardEffect | InteractiveCardEffect;

export const initialBasicCardEffect: BasicCardEffect = {
  kind: CardEffectKind.Basic,
  category: CardEffectCategory.Draw,
  value: 0,
};

export const initialInteractiveCardEffect: InteractiveCardEffect = {
  kind: CardEffectKind.Interactive,
  category: InteractiveCardEffectCategory.Discard,
  resolveType: InteractiveCardEffectResolveType.Optional,
  numPlaysToResolve: 0,
};
