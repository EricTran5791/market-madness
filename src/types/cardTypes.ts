import {
  CardEffect,
  CardEffectKind,
  CardEffectCategory,
  InteractiveCardEffectCategory,
  InteractiveCardEffectResolveType,
} from './cardEffect.types';

export enum CardCategory {
  Action = 'Action',
  Consumable = 'Consumable',
  Item = 'Item',
  Money = 'Money',
  NPC = 'NPC',
}

export enum CardSubcategory {
  Food = 'Food',
  SportsEquipment = 'Sports Equipment',
  Tool = 'Tool',
}

export enum CardCostKind {
  Money = 'Money',
  Health = 'Health',
}

export type Card = {
  name: string;
  category: CardCategory;
  subcategories: CardSubcategory[];
  description: string;
  cost: {
    kind: CardCostKind;
    value: number;
  };
  effects: CardEffect[];
  isPlayed?: boolean;
};

export const initialCard: Card = {
  name: '',
  category: CardCategory.Item,
  subcategories: [],
  description: '',
  cost: {
    kind: CardCostKind.Money,
    value: 0,
  },
  // TEMP
  effects: [
    {
      category: CardEffectCategory.GainAttack,
      kind: CardEffectKind.Basic,
      value: 2,
    },
    {
      category: CardEffectCategory.Heal,
      kind: CardEffectKind.Basic,
      value: 2,
    },
    {
      category: InteractiveCardEffectCategory.Trash,
      kind: CardEffectKind.Interactive,
      resolveType: InteractiveCardEffectResolveType.Mandatory,
      numCardsToResolve: 2,
    },
  ],
};
