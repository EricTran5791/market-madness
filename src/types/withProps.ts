import { ThemedStyledFunction } from 'styled-components';

/** This is an unfortunate workaround to provide props to styled elements in Typescript:
 *  https://github.com/styled-components/styled-components/issues/630
 */
export const withProps = <U>() => <P, T, O>(
  fn: ThemedStyledFunction<P, T, O>
) => fn as ThemedStyledFunction<P & U, T, O & U>;
