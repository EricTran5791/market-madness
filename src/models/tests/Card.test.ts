import { Card, CardModelType, CardStack, CardStackModelType } from '../Card';
import { getSnapshot } from 'mobx-state-tree';
import { initialCardState } from '../../types/cardTypes';

describe('Card and CardStack Model', () => {
  let mockCard: CardModelType;
  let mockCardStack: CardStackModelType;

  beforeEach(() => {
    mockCard = Card.create({
      uniqid: 'mock-id-0',
      ...initialCardState,
      name: 'Mock Card 0',
    });
    mockCardStack = CardStack.create({
      cards: [mockCard],
    });
  });

  it('creates a Card', () => {
    expect(getSnapshot(mockCard)).toMatchSnapshot();
  });

  it('creates a CardStack', () => {
    expect(getSnapshot(mockCardStack)).toMatchSnapshot();
  });

  it('adds a card to the CardStack', () => {
    mockCardStack.add(
      Card.create({
        uniqid: 'mock-id-1',
        ...initialCardState,
        name: 'Mock Card 1',
      })
    );
    expect(getSnapshot(mockCardStack)).toMatchSnapshot();
  });

  it('removes a card from the CardStack', () => {
    mockCardStack.remove(mockCard);
    expect(getSnapshot(mockCardStack)).toMatchSnapshot();
  });
});
