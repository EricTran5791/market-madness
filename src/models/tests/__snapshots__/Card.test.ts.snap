// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card and CardStack Model adds a card to the CardStack 1`] = `
Object {
  "cards": Array [
    Object {
      "categories": Array [],
      "cost": Object {
        "kind": "Money",
        "value": 0,
      },
      "description": "",
      "effects": Array [],
      "id": "1",
      "isPlayed": false,
      "kind": "Normal",
      "name": "Card 1",
      "shop": "Neutral",
      "uniqid": "1",
    },
    Object {
      "categories": Array [],
      "cost": Object {
        "kind": "Money",
        "value": 0,
      },
      "description": "",
      "effects": Array [],
      "id": "2",
      "isPlayed": false,
      "kind": "Normal",
      "name": "Card 2",
      "shop": "Neutral",
      "uniqid": "2",
    },
  ],
}
`;

exports[`Card and CardStack Model creates a Card 1`] = `
Object {
  "categories": Array [],
  "cost": Object {
    "kind": "Money",
    "value": 0,
  },
  "description": "",
  "effects": Array [],
  "id": "1",
  "isPlayed": false,
  "kind": "Normal",
  "name": "Card 1",
  "shop": "Neutral",
  "uniqid": "1",
}
`;

exports[`Card and CardStack Model creates a CardStack 1`] = `
Object {
  "cards": Array [
    Object {
      "categories": Array [],
      "cost": Object {
        "kind": "Money",
        "value": 0,
      },
      "description": "",
      "effects": Array [],
      "id": "1",
      "isPlayed": false,
      "kind": "Normal",
      "name": "Card 1",
      "shop": "Neutral",
      "uniqid": "1",
    },
  ],
}
`;

exports[`Card and CardStack Model removes a card from the CardStack 1`] = `
Object {
  "cards": Array [],
}
`;
